java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java

```diff
@@ -167,133 +124,222 @@ private static List<String> getPkColumnNames(
         return pkColumnNames;
     }
 
-    private PreparedStatement prepareInsertStatement(SinkRow row) {
-        if (row.getOp() != Data.Op.INSERT) {
-            throw Status.FAILED_PRECONDITION
-                    .withDescription("unexpected op type: " + row.getOp())
-                    .asRuntimeException();
-        }
-        try {
-            var preparedStmt = insertPreparedStmt;
-            jdbcDialect.bindInsertIntoStatement(preparedStmt, conn, getTableSchema(), row);
-            preparedStmt.addBatch();
-            return preparedStmt;
-        } catch (SQLException e) {
-            throw io.grpc.Status.INTERNAL
-                    .withDescription(
-                            String.format(ERROR_REPORT_TEMPLATE, e.getSQLState(), e.getMessage()))
-                    .withCause(e)
-                    .asRuntimeException();
-        }
-    }
+    @Override
+    public void write(Iterable<SinkRow> rows) {
+        int retryCount = 0;
+        while (true) {
+            try {
+                // fill prepare statements with parameters
+                for (SinkRow row : rows) {
+                    if (row.getOp() == Data.Op.UPDATE_DELETE) {
+                        updateFlag = true;
+                        continue;
+                    }
+                    if (config.isUpsertSink()) {
+                        if (row.getOp() == Data.Op.DELETE) {
+                            jdbcStatements.prepareDelete(row);
+                        } else {
+                            jdbcStatements.prepareUpsert(row);
+                        }
+                    } else {
+                        jdbcStatements.prepareInsert(row);
+                    }
+                }
+                // Execute staging statements after all rows are prepared.
+                jdbcStatements.execute();
+                break;
+            } catch (SQLException e) {
+                LOG.error("Failed to execute JDBC statements, retried {} times", retryCount, e);
```

Comment:
Seems that we are endlessly retrying? Shall we set a max retry count? And is there a way to check whether the SQLException is caused by network connection or some other data failure (such key len exceed limit)? It's better to create the connection when the exception is caused by network error.